### PR TITLE
MULE-11083: artifact plugin descriptor factory now detects if there's…

### PR DESCRIPTION
**Pull request by fernandezlautaro**
… a JSON file when loading a plugin. If it does exists, it will consume just the JSON file (not taking the plugin.properties for any loading

**## Do not merge until CB plan goes green here**: https://muleesb.ci.cloudbees.com/job/MULE-LAUTARO/32/

This PR looks reading the JSON mule-plugin.json file when initializing `ArtifactPluginDescriptor`s rather than the plugin.propeties file.

### CODE REVIEW

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java [hunk 1]
```diff
@@ -77,50 +109,38 @@ public ArtifactPluginDescriptor create(File pluginFolder) throws ArtifactDescrip
...
-    try (InputStream stream = new FileInputStream(jsonFile)) {
-      return new MulePluginModelJsonSerializer().deserialize(IOUtils.toString(stream));
-    } catch (IOException e) {
-      throw new IllegalArgumentException(format("Could not read extension describer on plugin '%s'", jsonFile.getAbsolutePath()),
-                                         e);
-    }
   }
 
+  /**
+   * // TODO(fernandezlautaro): MULE-11092 once implemented, drop this.
```
Comment (pablokraan) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: Don't add a TODO for each thing that must be removed, detecting what has to be cleaned up is part of the work once the loadFromPluginProperties method is removed

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java [hunk 2]
```diff
@@ -152,16 +175,140 @@ private BundleDependency createBundleDescriptor(String bundle) {
...
+      loadClassLoaderModelBuilderFromDescriptor(pluginFolder, descriptor, classLoaderModelBuilder, classLoaderModel);
+    });
+
+    return classLoaderModelBuilder.build();
+  }
+
+  private void loadClassLoaderModelBuilderFromDescriptor(File pluginFolder, ArtifactPluginDescriptor descriptor,
+                                                         ClassLoaderModelBuilder classLoaderModelBuilder,
+                                                         MulePluginLoaderDescriptor classLoaderModel) {
+    if (!classLoaderModel.getId().equals(MAVEN_ID_CLASSLOADER_MODEL_DESCRIPTOR)) {
```
Comment (pablokraan) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: Add a TODO to dynamically register loaders

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java [hunk 3]
```diff
@@ -152,16 +175,140 @@ private BundleDependency createBundleDescriptor(String bundle) {
...
+    classLoaderModelBuilder
+        .exportingPackages(new HashSet<>(getAttribute(classLoaderModel, EXPORTED_PACKAGES)))
+        .exportingResources(new HashSet<>(getAttribute(classLoaderModel, EXPORTED_RESOURCES)));
+
+
+    Model model = getPomModel(pluginFolder, mulePluginPom);
+    final Set<BundleDependency> plugins = model.getDependencies().stream()
+        .filter(dependency -> isMulePlugin(dependency))
+        .map(dependency -> {
+          final BundleDescriptor.Builder bundleDescriptorBuilder = new BundleDescriptor.Builder()
```
Comment (pablokraan) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: import for BundleDescriptor.Builder?

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java [hunk 4]
```diff
@@ -152,16 +175,140 @@ private BundleDependency createBundleDescriptor(String bundle) {
...
+        .setGroupId(model.getGroupId())
+        .setVersion(model.getVersion())
+        .setType(EXTENSION_BUNDLE_TYPE)
+        .setClassifier(MULE_PLUGIN_CLASSIFIER)
+        .build();
+    descriptor.setBundleDescriptor(bundleDescriptor);
+  }
+
+  private boolean isMulePlugin(Dependency dependency) {
+    // We will keep only Mule plugins dependencies or org.mule.runtime.deployment.model.internal.plugin.BundlePluginDependenciesResolver.getArtifactsWithDependencies() will fail looking them up.
```
Comment (pablokraan) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: Is this something that has to be fixed on a separated PR?

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java [hunk 5]
```diff
@@ -1539,6 +1542,85 @@ public void deploysAppZipWithExtensionXmlPlugin() throws Exception {
...
+        .containingMetaInfResource("module-bye-pom-with-mule-plugin-dependencies.xml", "pom.xml")
+        .describedBy(builder.build());
+
+    ApplicationFileBuilder applicationFileBuilder = new ApplicationFileBuilder("appWithExtensionXmlPlugin")
+        .definedBy("app-with-extension-xml-plugin-config.xml").containingPlugin(byeXmlExtensionPlugin);
+    addPackedAppFromBuilder(applicationFileBuilder);
+
+    startDeployment();
+
+    // TODO(fernandezlautaro): MULE-11089 the following line is expecting to see two deploys when it MUST be just one, for some reason tries to redeploy two times so I added until this bug gets fixed
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: the following line is expecting to see two deploys when it MUST be just one >the following line is expecting to see one deploy

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java [hunk 6]
```diff
@@ -1559,6 +1641,27 @@ public void failsToDeployWithExtensionThatHasNonExistingIdForExtensionModel() th
...
+    final ArtifactPluginFileBuilder byeXmlExtensionPlugin = new ArtifactPluginFileBuilder(extensionName)
+        .describedBy(builder.build());
+
+    ApplicationFileBuilder applicationFileBuilder = new ApplicationFileBuilder("appWithExtensionXmlPluginFails")
+        .definedBy("app-with-extension-xml-plugin-config.xml").containingPlugin(byeXmlExtensionPlugin);
+    addPackedAppFromBuilder(applicationFileBuilder);
+
+    startDeployment();
+
+    // TODO(fernandezlautaro): MULE-11089 the following line is expecting to see two deploys when it MUST be just one, for some reason tries to redeploy two times so I added until this bug gets fixed
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: Same as before

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java [hunk 7]
```diff
@@ -1745,6 +1848,22 @@ public void failsToDeployApplicationWithMissingPluginDependencyOnPlugin() throws
...
+        new ArtifactPluginFileBuilder("dependantPlugin")
+            .dependingOn(echoPlugin.getId());
+
+    final TestArtifactDescriptor artifactFileBuilder = new ApplicationFileBuilder("plugin-depending-on-plugin-app")
+        .definedBy("plugin-depending-on-plugin-app-config.xml").containingPlugin(dependantPlugin);
+    addPackedAppFromBuilder(artifactFileBuilder);
+
+    startDeployment();
+
+    // TODO(fernandezlautaro): MULE-11089 the following line is expecting to see two deploys when it MUST be just one, for some reason tries to redeploy two times so I added until this bug gets fixed
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: Again

File: modules/deployment/src/test/resources/module-bye-pom-with-mule-plugin-dependencies.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            <artifactId>mule-module-artifact</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>test</groupId>
+            <artifactId>echoPlugin</artifactId>
+            <version>1.0</version>
+            <classifier>mule-plugin</classifier>
+            <scope>provided</scope>
+            <!--BundleDescriptor{groupId='test', artifactId='echoPlugin', version='1.0', type='zip', classifier=Optional[mule-plugin]}-->
```
Comment (pablokraan) [inline modules/deployment/src/test/resources/module-bye-pom-with-mule-plugin-dependencies.xml]: Remove

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java [hunk 9]
```diff
@@ -1539,6 +1542,85 @@ public void deploysAppZipWithExtensionXmlPlugin() throws Exception {
   }
 
   @Test
+  public void deploysWithExtensionXmlPluginWithPomWithoutMulePluginDependencies() throws Exception {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: I think that only one test must cover deployment of plugins using the json, the other ones look like tests on ArtifactPluginDescriptorFactory, once you test that the created descriptor is the right one, there is no need to deploy a full application

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java (hunk 4 above)
Comment (fernandezlautaro) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: it's how the BundlePluginDependenciesResolver works, which I'm not sure if we should discriminate those deps that are not mule plugins here, or in the tBundlePluginDependenciesResolve class.

I can add a jira to discuss it later

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java (hunk 1 above)
Comment (fernandezlautaro) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: done.

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java (hunk 2 above)
Comment (fernandezlautaro) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: done

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java (hunk 3 above)
Comment (fernandezlautaro) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: done

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java (hunk 5 above)
Comment (fernandezlautaro) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: done

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java (hunk 6 above)
Comment (fernandezlautaro) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: done

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java (hunk 7 above)
Comment (fernandezlautaro) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/DeploymentServiceTestCase.java]: done

File: modules/deployment/src/test/resources/module-bye-pom-with-mule-plugin-dependencies.xml (hunk 8 above)
Comment (fernandezlautaro) [inline modules/deployment/src/test/resources/module-bye-pom-with-mule-plugin-dependencies.xml]: done

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java (hunk 4 above)
Comment (fernandezlautaro) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/plugin/ArtifactPluginDescriptorFactory.java]: added a TODO to discuss it later on next week.

**Pull request merged**